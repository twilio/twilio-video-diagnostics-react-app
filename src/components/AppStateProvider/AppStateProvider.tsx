import React, { createContext, useContext, useReducer, useEffect } from 'react';
import produce from 'immer';
import { VideoInputTest } from '@twilio/rtc-diagnostics';

export enum ActivePane {
  GetStarted,
  DeviceCheck,
  DeviceError,
  CameraTest,
  Connectivity,
  Quality,
  Results,
}

interface stateType {
  activePane: ActivePane;
  videoGranted: boolean;
  audioGranted: boolean;
  deviceError: null | Error;
  videoInputTestReport: null | VideoInputTest.Report;
}

type ACTIONTYPE =
  | { type: 'set-active-pane'; newActivePane: ActivePane }
  | { type: 'next-pane' }
  | { type: 'previous-pane' }
  | { type: 'set-devices'; devices: MediaDeviceInfo[] }
  | { type: 'set-device-error'; error: Error }
  | { type: 'set-video-test-report'; report: VideoInputTest.Report };

type AppStateContextType = {
  state: stateType;
  dispatch: React.Dispatch<ACTIONTYPE>;
};

export const initialState = {
  activePane: ActivePane.GetStarted,
  videoGranted: false,
  audioGranted: false,
  deviceError: null,
  videoInputTestReport: null,
};

export const AppStateContext = createContext<AppStateContextType>(null!);

export function useAppStateContext() {
  const context = useContext(AppStateContext);
  if (!context) {
    throw new Error('useAppStateContext must be used within an AppStateProvider');
  }
  return context;
}

/**
 * This reducer function helps with the logic used to decide the activePane,
 * as well as which panes are to be shown to the user, and when.
 */

export const appStateReducer = produce((draft: stateType, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'set-active-pane':
      draft.activePane = action.newActivePane;
      break;

    case 'next-pane':
      switch (draft.activePane) {
        case ActivePane.GetStarted:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.CameraTest;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        case ActivePane.DeviceCheck:
          if (draft.deviceError) {
            draft.activePane = ActivePane.DeviceError;
          } else {
            draft.activePane = ActivePane.CameraTest;
          }
          break;
        default:
          draft.activePane++;
          break;
      }
      break;

    case 'previous-pane':
      switch (draft.activePane) {
        case ActivePane.CameraTest:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.GetStarted;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        default:
          draft.activePane--;
          break;
      }
      break;

    case 'set-devices':
      draft.audioGranted = action.devices.filter((d) => d.kind === 'audioinput').every((d) => d.label);
      draft.videoGranted = action.devices.filter((d) => d.kind === 'videoinput').every((d) => d.label);
      break;

    case 'set-device-error':
      draft.deviceError = action.error;
      draft.activePane = ActivePane.DeviceError;
      break;

    case 'set-video-test-report':
      draft.videoInputTestReport = action.report;
  }
});

export const AppStateProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(appStateReducer, initialState);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then((devices) => dispatch({ type: 'set-devices', devices }));
  }, []);

  return <AppStateContext.Provider value={{ state, dispatch }}>{children}</AppStateContext.Provider>;
};
