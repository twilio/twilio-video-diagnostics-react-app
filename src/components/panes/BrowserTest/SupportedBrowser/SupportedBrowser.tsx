import UAParser from 'ua-parser-js';
import { makeStyles, Button, Container, Grid, Typography, Paper } from '@material-ui/core';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';

const useStyles = makeStyles({
  paper: {
    padding: '1.5em',
    borderRadius: '8px',
    width: '388px',
  },
  paperContainer: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
});

export function SupportedBrowser() {
  const { dispatch } = useAppStateContext();
  const classes = useStyles();

  const userAgentParser = new UAParser();
  const browser = userAgentParser.getBrowser();
  const operatingSystem = userAgentParser.getOS();

  return (
    <>
      <Container>
        <Grid container alignItems="center" justifyContent="space-between">
          <Grid item md={5}>
            <Typography variant="h1" gutterBottom>
              Browser supported
            </Typography>
            <Typography variant="body1" gutterBottom>
              Looks like your browser is good to go. We're almost done!
            </Typography>
            <Button
              variant="contained"
              color="primary"
              onClick={() => dispatch({ type: 'next-pane' })}
              style={{ marginRight: '1.5em' }}
            >
              Ok
            </Button>
          </Grid>
          ​
          <Grid item md={5} className={classes.paperContainer}>
            <Paper className={classes.paper}>
              <Typography variant="body1">
                <strong>Browser: </strong>
                {browser.name} {browser.version}
              </Typography>
              ​
              <Typography variant="body1">
                <strong>Operating System: </strong>
                {operatingSystem.name} {operatingSystem.version}
              </Typography>
            </Paper>
          </Grid>
        </Grid>
      </Container>
    </>
  );
}
