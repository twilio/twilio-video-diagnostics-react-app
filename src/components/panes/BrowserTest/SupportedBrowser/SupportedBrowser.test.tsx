import { shallow } from 'enzyme';
import { Button } from '@material-ui/core';
import { SupportedBrowser } from './SupportedBrowser';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';

Object.defineProperty(navigator, 'userAgent', {
  value:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36',
});

jest.mock('../../../AppStateProvider/AppStateProvider');

const mockUseAppStateContext = useAppStateContext as jest.Mock<any>;
const mockDispatch = jest.fn();

mockUseAppStateContext.mockImplementation(() => ({
  dispatch: mockDispatch,
}));

describe('the Supported Browser component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<SupportedBrowser />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should go to the next pane when the "Ok" button is clicked', () => {
    const wrapper = shallow(<SupportedBrowser />);

    wrapper.find(Button).simulate('click');
    expect(mockDispatch).toHaveBeenCalledWith({ type: 'next-pane' });
  });
});
