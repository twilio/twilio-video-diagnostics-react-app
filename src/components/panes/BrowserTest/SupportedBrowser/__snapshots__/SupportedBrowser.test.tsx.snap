// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Supported Browser component should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Container))>
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      container={true}
      justifyContent="space-between"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        md={5}
      >
        <WithStyles(ForwardRef(Typography))
          gutterBottom={true}
          variant="h1"
        >
          Browser supported
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          gutterBottom={true}
          variant="body1"
        >
          Looks like your browser is good to go. We're almost done!
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Button))
          color="primary"
          onClick={[Function]}
          style={
            Object {
              "marginRight": "1.5em",
            }
          }
          variant="contained"
        >
          Ok
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-paperContainer-2"
        item={true}
        md={5}
      >
        <WithStyles(ForwardRef(Paper))
          className="makeStyles-paper-1"
        >
          <WithStyles(ForwardRef(Typography))
            variant="body1"
          >
            <strong>
              Browser: 
            </strong>
            Chrome
             
            92.0.4515.131
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            variant="body1"
          >
            <strong>
              Operating System: 
            </strong>
            Mac OS
             
            10.15.7
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Paper))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Container))>
</Fragment>
`;
